Convert shader JIT to Xbyak

yuriks opened the pull request:
~Currently this crashes because the Citra and Dynarmic versions of Xybak conflict and cause ODR violations. Use merrymage/dynarmic#62 to fix this.~ Since merged and added to the PR.

Unified diff is probably more pleasant than Split for this one. I tried to keep this code a straight semantic port from the previous one. I'll make some cleanups in following PRs.

lioncash commented on src/common/x64/xbyak_abi.h:
```diff
@@ -0,0 +1,213 @@
+inline Xbyak::Reg IndexToReg(int reg_index) {
+    if (reg_index < 16) {
+        return IndexToReg64(reg_index);
+    } else {
```
else isn't necessary here, since the bodies are stacked returns.

yuriks replied:
I like the symmetry of the else here. There's two cases and they're both equally important.

lioncash replied:
I don't really understand the notion that not including redundant control specifiers means a particular branch is less important, but OK.

I get the symmetry one though, I suppose.

yuriks commented:
Xbyak isn't a module because I was originally going to use a different file structure where it'd get it's own include directory (so you'd use paths like <xbyak/xbyak.h>). But then I realized all files have a xbyak_ prefix anyway, so that would be kinda pointless. I'll change it.

lioncash commented on src/common/x64/xbyak_util.h:
```diff
@@ -0,0 +1,49 @@
+
+template <typename T>
+inline void CallFarFunction(Xbyak::CodeGenerator& code, const T f) {
+    static_assert(std::is_pointer<T>(), "Argument must be a (function) pointer.");
```
I missed this when initially reviewing the first time, but you can do:

```
std::is_pointer<T>() && std::is_function<std::remove_pointer_t<T>>()
```

as your static assert to ensure the given type is a function pointer. If the JIT just uses opaque pointers to functions (ew), disregard this.

yuriks replied:
I didn't add that other check in case the caller *does* have an opaque pointer (which is not our case but I could totally see it happening), so they can call this without having to cast it to some dummy function type. (Since this function doesn't care about the signature or calling convention of the pointer at all.)

yuriks commented on src/video_core/shader/shader_jit_x64.h:
```diff
@@ -103,17 +104,16 @@ class JitShader : public Gen::XCodeBlock {
     void FindReturnOffsets();
 
     /// Mapping of Pica VS instructions to pointers in the emitted code
+    // TODO(yuriks): instruction_labels already keeps track of this, but xbyak doesn't let us it
```
Fixed

yuriks commented:
Updated:
 - Small cleanup with the register types
 - Fixed formatting errors
 - Made xbyak a git submodule
 - Use `-fno-operator-names` on non-MSVC

bunnei commented:
LGTM, other than the build issues ofc on GCC/clang

yuriks commented:
Fixed build. Also, since herumi added a feature I requested (herumi/xbyak#43) I've taken care of that TODO, if anyone cares enough to review it. Otherwise, I'll merge this tomorrow.

bunnei commented:
Reviewed the recent changes. Still LGTM

MerryMage commented:
Looks fine to me.
